packages/core/data-transfer/src/engine/diagnostic.ts

```diff
@@ -0,0 +1,119 @@
+import { EventEmitter } from 'events';
+
+export interface IDiagnosticReporterOptions {
+  stackSize?: number;
+}
+
+export type GenericDiagnostic<K extends DiagnosticKind, T = unknown> = {
+  kind: K;
+  details: {
+    message: string;
+    createdAt: Date;
+  } & T;
+};
+
+export type DiagnosticKind = 'error' | 'warning' | 'info';
+
+export type DiagnosticListener<T extends DiagnosticKind = DiagnosticKind> = (
+  diagnostic: { kind: T } & Diagnostic extends infer U ? U : 'foo'
+) => void | Promise<void>;
+
+export type DiagnosticEvent = 'diagnostic' | `diagnostic.${DiagnosticKind}`;
+
+export type GetEventListener<E extends DiagnosticEvent> = E extends 'diagnostic'
+  ? DiagnosticListener
+  : E extends `diagnostic.${infer K}`
+  ? K extends DiagnosticKind
+    ? DiagnosticListener<K>
+    : never
+  : never;
+
+export type Diagnostic = ErrorDiagnostic | WarningDiagnostic | InfoDiagnostic;
+
+export type ErrorDiagnosticSeverity = 'fatal' | 'error' | 'silly';
+
+export type ErrorDiagnostic = GenericDiagnostic<
+  'error',
+  {
+    name: string;
+    severity: ErrorDiagnosticSeverity;
+    error: Error;
+  }
+>;
+
+export type WarningDiagnostic = GenericDiagnostic<
+  'warning',
+  {
+    origin?: string;
+  }
+>;
+
+export type InfoDiagnostic<T = unknown> = GenericDiagnostic<
```

Comment:
What about having levels of info like error has? There is some info that is important (like "transfer complete") and some that would be used with --verbose, and some that would be --debug level (like "entity from provider... to transform... to provider... finished")